indexer/displacement_manager.hpp

```diff
@@ -0,0 +1,226 @@
+
+  CellFeaturePair(uint64_t cell, uint32_t feature)
+    : m_cellLo(UINT64_LO(cell)), m_cellHi(UINT64_HI(cell)), m_feature(feature)
+  {
+  }
+
+  bool operator<(CellFeaturePair const & rhs) const
+  {
+    if (m_cellHi != rhs.m_cellHi)
+      return m_cellHi < rhs.m_cellHi;
+    if (m_cellLo != rhs.m_cellLo)
+      return m_cellLo < rhs.m_cellLo;
+    return m_feature < rhs.m_feature;
+  }
+
+  uint64_t GetCell() const { return UINT64_FROM_UINT32(m_cellHi, m_cellLo); }
+  uint32_t GetFeature() const { return m_feature; }
+private:
+  uint32_t m_cellLo;
+  uint32_t m_cellHi;
+  uint32_t m_feature;
+};
+static_assert(sizeof(CellFeaturePair) == 12, "");
+#ifndef OMIM_OS_LINUX
+static_assert(is_trivially_copyable<CellFeaturePair>::value, "");
+#endif
+
+class CellFeatureBucketTuple
+{
+public:
+  CellFeatureBucketTuple() : m_bucket(0) {}
+  CellFeatureBucketTuple(CellFeaturePair const & p, uint32_t bucket) : m_pair(p), m_bucket(bucket)
+  {
+  }
+
+  bool operator<(CellFeatureBucketTuple const & rhs) const
+  {
+    if (m_bucket != rhs.m_bucket)
+      return m_bucket < rhs.m_bucket;
+    return m_pair < rhs.m_pair;
+  }
+
+  CellFeaturePair const & GetCellFeaturePair() const { return m_pair; }
+  uint32_t GetBucket() const { return m_bucket; }
+private:
+  CellFeaturePair m_pair;
+  uint32_t m_bucket;
+};
+static_assert(sizeof(CellFeatureBucketTuple) == 16, "");
+#ifndef OMIM_OS_LINUX
+static_assert(is_trivially_copyable<CellFeatureBucketTuple>::value, "");
+#endif
+
+/// Displacement manager filters incoming single-point features to simplify runtime
+/// feature visibility displacement.
+template <class TSorter>
+class DisplacementManager
+{
+public:
+  DisplacementManager(TSorter & sorter) : m_sorter(sorter) {}
+
+  /// Add feature at bucket (zoom) to displacable queue if possible. Pass to bucket otherwise.
+  template <class TFeature>
+  void Add(vector<int64_t> const & cells, uint32_t bucket, TFeature const & ft, uint32_t index)
+  {
+    if (bucket != scales::GetUpperScale() && IsDisplaceable(ft))
+    {
+      m_storage[bucket].emplace_back(cells, ft, index, bucket);
+      return;
+    }
+
+    for (auto const & cell : cells)
+      m_sorter.Add(CellFeatureBucketTuple(CellFeaturePair(cell, index), bucket));
+  }
+
+  /// Check features intersection and pass result to sorter.
+  void Displace()
+  {
+    m4::Tree<DisplaceableNode> acceptedNodes;
+    list<DisplaceableNode> deferredNodes;
+    for (uint32_t scale = 0; scale < scales::GetUpperScale(); ++scale)
+    {
+      // Initialize queue.
+      priority_queue<DisplaceableNode> displaceableQueue;
+      for (auto const & node : m_storage[scale])
+        displaceableQueue.push(node);
+      for (auto const & node : deferredNodes)
+        displaceableQueue.push(node);
+      deferredNodes.clear();
+
+      float const delta = CalculateDeltaForZoom(scale);
+      size_t accepted = 0;
+
+      // Process queue.
+      while (!displaceableQueue.empty())
```

Comment:
You're adding to the queue all elements at once and then just extract them - why not to use plain vector + std::sort?